File: problems/0017-letter-combinations-of-a-phone-number/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>

class Solution {
private:
    std::vector<std::string> combinations;  // Stores the final result.
    std::unordered_map<char, std::string> letters = {
        {'2', "abc"}, {'3', "def"},  {'4', "ghi"}, {'5', "jkl"},
        {'6', "mno"}, {'7', "pqrs"}, {'8', "tuv"}, {'9', "wxyz"}
    };
    std::string phoneDigits;  // Stores the original input digits for recursive access.

public:
    std::vector<std::string> letterCombinations(std::string digits) {
        /*
        Problem 0017 – Letter Combinations of a Phone Number.

        Given a string containing digits from 2 to 9, return all possible letter combinations
        that the number could represent, based on the telephone keypad.

        Return the answer in any order.

        Time Complexity  : O(4^n × n), where n is the length of the input string.
                           Each digit contributes up to 4 letters, and we build strings of length n.
        Space Complexity : O(n) recursion depth, or O(4^n × n) including output.
        */

        if (digits.empty()) {
            return combinations;  // Return empty if no digits are provided.
        }

        phoneDigits = digits;
        std::string path;  // Holds current path during backtracking.
        backtrack(0, path);
        return combinations;
    }

private:
    void backtrack(int index, std::string& path) {
        // If we've constructed a full combination, add it to the result.
        if (path.length() == phoneDigits.length()) {
            combinations.push_back(path);
            return;
        }

        // Get the possible letters corresponding to the current digit.
        const std::string& possibleLetters = letters[phoneDigits[index]];

        for (char letter : possibleLetters) {
            path.push_back(letter);            // Choose a letter.
            backtrack(index + 1, path);        // Recurse to the next digit.
            path.pop_back();                   // Backtrack by removing the last letter.
        }
    }
};

File: problems/0017-letter-combinations-of-a-phone-number/solution.py
class Solution:
    def letterCombinations(self, digits: str) -> list[str]:
        """
        Problem 0017 – Letter Combinations of a Phone Number.

        Given a string containing digits from 2 to 9, return all possible letter combinations
        that the number could represent, based on the telephone keypad.

        Return the answer in any order.

        Time Complexity  : O(4^n × n), where n is the length of the input string.
                           Each digit can contribute up to 4 letters, and we build strings of length n.
        Space Complexity : O(n) for recursion, or O(4^n × n) if including output.
        """

        if len(digits) == 0:
            return []

        letters = {
            "2": "abc", "3": "def", "4": "ghi", "5": "jkl",
            "6": "mno", "7": "pqrs", "8": "tuv", "9": "wxyz"
        }

        def backtrack(idx, path):
            # We have completed a valid combination.
            if len(path) == len(digits):
                combinations.append("".join(path))
                return

            possible_letters = letters[digits[idx]]
            for letter in possible_letters:
                path.append(letter)        # Choose a letter.
                backtrack(idx + 1, path)   # Explore deeper.
                path.pop()                 # Backtrack to try next letter.

        combinations = []
        backtrack(0, [])

        return combinations

